File: Chap6-GraphicsViewFramework/6-14PainterAppRubberBand/resizablehandlerect.py
from PySide6.QtCore import QRectF, QSize, QPointF
from PySide6.QtGui import QPen, QBrush
from PySide6.QtWidgets import QGraphicsItem
from PySide6.QtCore import Qt

class ResizableHandleRect:
    """Base class for resizable items with handles at the corners"""
    
    def __init__(self):
        self.topleft_handle_rect = QRectF()
        self.top_right_handle_rect = QRectF()
        self.bottom_right_handle_rect = QRectF()
        self.bottom_left_handle_rect = QRectF()
        
        self.handle_list = []
        self.handles_added_to_scene = False
        self.owner_item = None
    
    def draw_handles(self):
        """Draw handles at the corners of the item"""
        from handleitem import HandleItem
        
        # Populate handles in list
        if len(self.handle_list) == 0:
            self.handle_list.append(HandleItem(HandleItem.TopLeft))
            self.handle_list.append(HandleItem(HandleItem.TopRight))
            self.handle_list.append(HandleItem(HandleItem.BottomRight))
            self.handle_list.append(HandleItem(HandleItem.BottomLeft))
        
        # Set up pen
        m_pen = QPen()
        m_pen.setWidth(2)
        m_pen.setColor(Qt.black)
        
        # Top left handle
        top_left_corner = self.selector_frame_bounds().topLeft() + QPointF(-12.0, -12.0)
        self.topleft_handle_rect = QRectF(top_left_corner, QSize(12, 12))
        self.handle_list[0].setRect(self.topleft_handle_rect)
        
        if len(self.handle_list) > 0 and not self.handles_added_to_scene:
            self.handle_list[0].setPen(m_pen)
            self.handle_list[0].setBrush(QBrush(Qt.gray))
            self.owner_item.scene().addItem(self.handle_list[0])
            self.handle_list[0].setParentItem(self.owner_item)
        
        # Top right
        top_right_corner = self.selector_frame_bounds().topRight() + QPointF(0, -12.0)
        self.top_right_handle_rect = QRectF(top_right_corner, QSize(12, 12))
        self.handle_list[1].setRect(self.top_right_handle_rect)
        
        if len(self.handle_list) > 0 and not self.handles_added_to_scene:
            self.handle_list[1].setPen(m_pen)
            self.handle_list[1].setBrush(QBrush(Qt.gray))
            self.owner_item.scene().addItem(self.handle_list[1])
            self.handle_list[1].setParentItem(self.owner_item)
        
        # Bottom right
        bottom_right_corner = self.selector_frame_bounds().bottomRight() + QPointF(0, 0)
        self.bottom_right_handle_rect = QRectF(bottom_right_corner, QSize(12, 12))
        self.handle_list[2].setRect(self.bottom_right_handle_rect)
        
        if len(self.handle_list) > 0 and not self.handles_added_to_scene:
            self.handle_list[2].setPen(m_pen)
            self.handle_list[2].setBrush(QBrush(Qt.gray))
            self.owner_item.scene().addItem(self.handle_list[2])
            self.handle_list[2].setParentItem(self.owner_item)
        
        # Bottom left
        bottom_left_corner = self.selector_frame_bounds().bottomLeft() + QPointF(-12, 0)
        self.bottom_left_handle_rect = QRectF(bottom_left_corner, QSize(12, 12))
        self.handle_list[3].setRect(self.bottom_left_handle_rect)
        
        if len(self.handle_list) > 0 and not self.handles_added_to_scene:
            self.handle_list[3].setPen(m_pen)
            self.handle_list[3].setBrush(QBrush(Qt.gray))
            self.owner_item.scene().addItem(self.handle_list[3])
            self.handle_list[3].setParentItem(self.owner_item)
        
        self.handles_added_to_scene = True
    
    def set_owner_item(self, item):
        """Set the owner item of this resizable handle"""
        self.owner_item = item
    
    def draw_handles_if_necessary(self):
        """Draw handles if the item is selected, remove them otherwise"""
        if not self.owner_item:
            import warnings
            warnings.warn("ResizableHandleRect: No ownerItem set. Not drawing any handle. "
                         "Please call set_owner_item on your ResizableHandleRect subclass")
            return
        
        if self.owner_item.isSelected():
            self.draw_handles()
            self.handles_added_to_scene = True
        else:
            # Remove the handles
            for handle_item in self.handle_list:
                self.owner_item.scene().removeItem(handle_item)
            
            # Clean up handles
            self.handle_list.clear()
            self.handles_added_to_scene = False
    
    # Abstract methods to be implemented by subclasses
    def selector_frame_bounds(self):
        """Returns the bounds of the selector frame"""
        raise NotImplementedError("Subclasses must implement selector_frame_bounds")
    
    def set_selector_frame_bounds(self, bounds_rect):
        """Sets the bounds of the selector frame"""
        raise NotImplementedError("Subclasses must implement set_selector_frame_bounds")

File: 17-Miscellaneous/7-4PainterAppSaveLoad/resizablehandlerect.cpp
#include "resizablehandlerect.h"
#include <QGraphicsScene>
#include <QDebug>

ResizableHandleRect::ResizableHandleRect()
{

}

ResizableHandleRect::~ResizableHandleRect()
{

}

void ResizableHandleRect::drawHandles()
{
    //Populate handles in list
    if(handleList.count() == 0){
        handleList.append(new HandleItem(HandleItem::TopLeft));
        handleList.append(new HandleItem(HandleItem::TopRight));
        handleList.append(new HandleItem(HandleItem::BottomRight));
        handleList.append(new HandleItem(HandleItem::BottomLeft));
    }


    //Set up pen
    QPen mPen;
    mPen.setWidth(2);
    mPen.setColor(Qt::black);

    //Top left handle
    QPointF topLeftCorner = selectorFrameBounds().topLeft() + QPointF(-12.0,-12.0);
    topleftHandleRect  = QRectF(topLeftCorner,QSize(12,12));
    handleList[0]->setRect(topleftHandleRect);
    if(!handleList.isEmpty() && (!handlesAddedToScene)){
        handleList[0]->setPen(mPen);
        handleList[0]->setBrush(QBrush(Qt::gray));
        ownerItem->scene()->addItem(handleList[0]);
        handleList[0]->setParentItem(ownerItem);

    }

    //Top right
    QPointF topRightCorner = selectorFrameBounds().topRight() + QPointF(0,-12.0);
    topRightHandleRect  = QRectF(topRightCorner,QSize(12,12));
    handleList[1]->setRect(topRightHandleRect);
    if(!handleList.isEmpty() && (!handlesAddedToScene)){
        handleList[1]->setPen(mPen);
        handleList[1]->setBrush(QBrush(Qt::gray));
        ownerItem->scene()->addItem(handleList[1]);
        handleList[1]->setParentItem(ownerItem);
    }

    //Bottom right
    QPointF bottomRightCorner = selectorFrameBounds().bottomRight() + QPointF(0,0);
    bottomRightHandleRect  = QRectF(bottomRightCorner,QSize(12,12));
    handleList[2]->setRect(bottomRightHandleRect);
    if(!handleList.isEmpty() && (!handlesAddedToScene)){
        handleList[2]->setPen(mPen);
        handleList[2]->setBrush(QBrush(Qt::gray));
        ownerItem->scene()->addItem(handleList[2]);
        handleList[2]->setParentItem(ownerItem);
    }

    //Bottom left
    QPointF bottomLeftCorner = selectorFrameBounds().bottomLeft() + QPointF(-12,0);
    bottomLeftHandleRect  = QRectF(bottomLeftCorner,QSize(12,12));
    handleList[3]->setRect(bottomLeftHandleRect);
    if(!handleList.isEmpty() && (!handlesAddedToScene)){
        handleList[3]->setPen(mPen);
        handleList[3]->setBrush(QBrush(Qt::gray));
        ownerItem->scene()->addItem(handleList[3]);
        handleList[3]->setParentItem(ownerItem);
    }

    handlesAddedToScene = true;

}

void ResizableHandleRect::setOwnerItem(QGraphicsItem *value)
{
    ownerItem = value;
}

void ResizableHandleRect::drawHandlesIfNecessary()
{
    if(!ownerItem){
        qWarning() << "ResizableHandleRect : No ownerItem set. Not drawing any\
                      handle. Please call setOwnerItem on your ResizableHandleRect subclass";
                      return;
    }



    if(ownerItem->isSelected()){
        drawHandles();
        handlesAddedToScene = true;
    }else{

        //Remove the handles
        foreach (HandleItem * handleItem, handleList) {
            ownerItem->scene()->removeItem(handleItem);
        }
        qDeleteAll(handleList);
        handleList.clear();
        handlesAddedToScene = false;
    }

}
